modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4MessageChannelHandler.java

```diff
@@ -68,4 +79,121 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
+            doFlush(ctx);
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        doFlush(ctx);
+        super.channelInactive(ctx);
+    }
+
+    private void resumeFlush() {
+        final ChannelHandlerContext context = this.ctx;
+        if (context == null) {
+            return;
+        }
+        if (context.executor().inEventLoop()) {
+            doFlush(context);
+        } else {
+            context.executor().execute(() -> doFlush(context));
+        }
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) {
+        this.ctx = ctx;
+    }
+
+    private void doFlush(ChannelHandlerContext ctx) {
+        final Channel channel = ctx.channel();
+        if (channel.isActive() == false) {
+            if (currentWrite != null) {
+                currentWrite.promise.tryFailure(new ClosedChannelException());
+            }
+            WriteOperation queuedWrite;
+            while ((queuedWrite = queuedWrites.poll()) != null) {
+                queuedWrite.promise.tryFailure(new ClosedChannelException());
+            }
+            return;
+        }
+        while (channel.isWritable()) {
+            if (currentWrite == null) {
+                currentWrite = queuedWrites.poll();
+            }
+            if (currentWrite == null) {
+                break;
+            }
+            final WriteOperation write = currentWrite;
+            if (write.buf.readableBytes() == 0) {
+                write.promise.trySuccess();
+                currentWrite = null;
+                continue;
+            }
+            final int bufferSize = Math.min(write.buf.readableBytes(), 1 << 18);
+            final ByteBuf writeBuffer = ctx.alloc().directBuffer(bufferSize, bufferSize);
+            writeBuffer.writeBytes(write.buf, bufferSize);
+            final ChannelFuture writeFuture = ctx.write(writeBuffer);
+            if (write.buf.readableBytes() == 0) {
+                currentWrite = null;
+                writeFuture.addListener((ChannelFutureListener) future -> {
+                    if (future.isSuccess()) {
+                        write.promise.trySuccess();
+                        resumeFlush();
+                    } else {
+                        write.promise.tryFailure(future.cause());
+                    }
+                });
+            } else {
+                writeFuture.addListener(f -> {
+                    if (f.isSuccess() == false) {
+                        write.promise.tryFailure(f.cause());
+                    } else {
+                        resumeFlush();
+                    }
+                });
+            }
+            ctx.flush();
+            if (channel.isActive() == false) {
+                WriteOperation queuedWrite;
+                while ((queuedWrite = queuedWrites.poll()) != null) {
+                    queuedWrite.promise.tryFailure(new ClosedChannelException());
```

Comment:
It is confusing that these are cleared in the `doFlush` call. Can we not just clear and release everything in the `channelInactive` callback? Or the `close` callback if you change this to an outbound handler and put it after the ssl handler?

Reply:
This solves the problem of potentially trying to flush released bytes because we can't just rely on the channel inactive hook as far as I can see because it triggers in a later iteration of the event-loop. From `io.netty.channel.AbstractChannel.AbstractUnsafe#close`:

```java
            // Fail all the queued messages
            try {
                outboundBuffer.failFlushed(CLOSED_CHANNEL_EXCEPTION);
                outboundBuffer.close(CLOSED_CHANNEL_EXCEPTION);
            } finally {

                if (wasActive && !isActive()) {
                    invokeLater(new OneTimeTask() {
                        @Override
                        public void run() {
                            pipeline.fireChannelInactive();
                        }
                    });
                }
```